File: bot-intelligence/src/main/java/gg/mineral/bot/ai/goal/ReplaceArmorGoal.java
package gg.mineral.bot.ai.goal;

import gg.mineral.bot.api.controls.Key;
import gg.mineral.bot.api.controls.MouseButton;

import gg.mineral.bot.api.event.Event;
import gg.mineral.bot.api.goal.Goal;

import gg.mineral.bot.api.instance.ClientInstance;
import gg.mineral.bot.api.inv.item.Item;

import gg.mineral.bot.api.screen.type.ContainerScreen;

import lombok.Getter;
import lombok.val;

@Getter
public class ReplaceArmorGoal extends Goal {

    private boolean inventoryOpen = false;

    @Override
    public boolean shouldExecute() {

        // TODO: don't replace armor if eating gapple or drinking potion

        val fakePlayer = clientInstance.getFakePlayer();
        val inventory = fakePlayer.getInventory();

        if (inventory == null)
            return false;

        val missingArmorPiece = missingArmorPiece();

        return canSeeEnemy() && missingArmorPiece != Item.Type.NONE && inventory.contains(missingArmorPiece);
    }

    public ReplaceArmorGoal(ClientInstance clientInstance) {
        super(clientInstance);
    }

    private Item.Type missingArmorPiece() {
        val fakePlayer = clientInstance.getFakePlayer();
        val inventory = fakePlayer.getInventory();
        if (inventory == null)
            return Item.Type.NONE;

        val helmet = inventory.getHelmet();
        val chestplate = inventory.getChestplate();
        val leggings = inventory.getLeggings();
        val boots = inventory.getBoots();

        if (helmet == null)
            return Item.Type.HELMET;
        if (chestplate == null)
            return Item.Type.CHESTPLATE;
        if (leggings == null)
            return Item.Type.LEGGINGS;

        return boots == null ? Item.Type.BOOTS : Item.Type.NONE;
    }

    private boolean canSeeEnemy() {
        val fakePlayer = clientInstance.getFakePlayer();
        val world = fakePlayer.getWorld();
        return world == null ? false
                : world.getEntities().stream()
                        .anyMatch(entity -> !clientInstance.getConfiguration().getFriendlyUUIDs()
                                .contains(entity.getUuid()));
    }

    private void applyArmor() {
        pressButton(10, MouseButton.Type.RIGHT_CLICK);
    }

    private void switchToArmor() {
        var armorSlot = -1;
        val fakePlayer = clientInstance.getFakePlayer();
        val inventory = fakePlayer.getInventory();

        if (inventory == null)
            return;

        val missingArmorPiece = missingArmorPiece();

        // Search hotbar
        for (int i = 0; i < 36; i++) {
            val itemStack = inventory.getItemStackAt(i);
            if (itemStack == null)
                continue;
            val item = itemStack.getItem();
            if (missingArmorPiece.isType(item.getId())) {
                armorSlot = i;
                break;
            }
        }

        if (armorSlot > 8) {
            if (!inventoryOpen) {
                inventoryOpen = true;
                pressKey(10, Key.Type.KEY_E);
                return;
            }

            val screen = clientInstance.getCurrentScreen();

            if (screen == null) {
                inventoryOpen = false;
                return;
            }
            // Move mouse
            val inventoryContainer = fakePlayer.getInventoryContainer();

            if (inventoryContainer == null) {
                inventoryOpen = false;
                return;
            }

            val slot = inventoryContainer.getSlot(inventory, armorSlot);

            if (slot == null) {
                inventoryOpen = false;
                pressKey(10, Key.Type.KEY_ESCAPE);
                return;
            }

            if (screen instanceof ContainerScreen containerScreen) {
                val x = containerScreen.getSlotX(slot);
                val y = containerScreen.getSlotY(slot);

                val currX = getMouseX();
                val currY = getMouseY();

                if (currX != x || currY != y) {
                    setMouseX(x);
                    setMouseY(y);
                } else {
                    // Move to end slot
                    pressKey(10, Key.Type.KEY_8);
                }

            }

            return;
        }

        if (inventoryOpen) {
            inventoryOpen = false;
            pressKey(10, Key.Type.KEY_ESCAPE);
            return;
        }

        pressKey(10, Key.Type.valueOf("KEY_" + (armorSlot + 1)));
    }

    @Override
    public void onTick() {
        val fakePlayer = clientInstance.getFakePlayer();
        val inventory = fakePlayer.getInventory();

        if (inventory == null)
            return;

        val itemStack = inventory.getHeldItemStack();

        val missingArmorPiece = missingArmorPiece();

        if (!delayedTasks.isEmpty())
            return;

        if (itemStack != null && missingArmorPiece.isType(itemStack.getItem().getId()))
            schedule(() -> applyArmor(), 100);
        else
            schedule(() -> switchToArmor(), 100);
    }

    @Override
    public boolean onEvent(Event event) {
        return false;
    }

    @Override
    public void onGameLoop() {
    }
}

File: bot-intelligence/src/main/java/gg/mineral/bot/ai/goal/ReplaceArmorGoal.kt
package gg.mineral.bot.ai.goal

import gg.mineral.bot.ai.goal.type.InventoryGoal
import gg.mineral.bot.api.controls.Key
import gg.mineral.bot.api.controls.MouseButton
import gg.mineral.bot.api.event.Event
import gg.mineral.bot.api.goal.Sporadic
import gg.mineral.bot.api.goal.Timebound
import gg.mineral.bot.api.instance.ClientInstance
import gg.mineral.bot.api.inv.item.Item
import gg.mineral.bot.api.screen.type.ContainerScreen

class ReplaceArmorGoal(clientInstance: ClientInstance) : InventoryGoal(clientInstance), Sporadic, Timebound {
    override val maxDuration: Long = 100
    override var startTime: Long = 0
    override var executing: Boolean = false

    override fun shouldExecute(): Boolean {
        val fakePlayer = clientInstance.fakePlayer
        val inventory = fakePlayer.inventory

        val missingArmorPiece = missingArmorPiece()

        return missingArmorPiece != Item.Type.NONE && inventory.contains(missingArmorPiece)
    }

    override fun onStart() {
    }

    private fun missingArmorPiece(): Item.Type {
        val fakePlayer = clientInstance.fakePlayer
        val inventory = fakePlayer.inventory

        val helmet = inventory.helmet
        val chestplate = inventory.chestplate
        val leggings = inventory.leggings
        val boots = inventory.boots

        if (helmet == null) return Item.Type.HELMET
        if (chestplate == null) return Item.Type.CHESTPLATE
        if (leggings == null) return Item.Type.LEGGINGS

        return if (boots == null) Item.Type.BOOTS else Item.Type.NONE
    }

    private fun getArmorSlot(): Int {
        var armorSlot = -1
        val fakePlayer = clientInstance.fakePlayer
        val inventory = fakePlayer.inventory

        val missingArmorPiece = missingArmorPiece()

        for (i in 0..35) {
            val itemStack = inventory.getItemStackAt(i) ?: continue
            val item = itemStack.item
            if (missingArmorPiece.isType(item.id)) {
                armorSlot = i
                break
            }
        }

        return armorSlot
    }

    override fun onTick(tick: Tick) {
        val armorSlot = getArmorSlot()
        val fakePlayer = clientInstance.fakePlayer
        val inventory = fakePlayer.inventory

        tick.finishIf("No Valid Armor Found", armorSlot == -1)

        tick.prerequisite("In Hotbar", armorSlot <= 8) {
            moveItemToHotbar(armorSlot, inventory)
        }

        tick.prerequisite("Inventory Closed", clientInstance.currentScreen !is ContainerScreen) {
            pressKey(10, Key.Type.KEY_ESCAPE)
        }

        tick.prerequisite("Correct Hotbar Slot Selected", inventory.heldSlot == armorSlot) {
            pressKey(10, Key.Type.valueOf("KEY_" + (armorSlot + 1)))
        }

        tick.finishIf(
            "Not Holding Valid Armor",
            inventory.heldItemStack?.let { missingArmorPiece().isType(it.item.id) } == false)

        tick.execute {
            pressButton(10, MouseButton.Type.RIGHT_CLICK)
        }
    }

    override fun onEnd() {
    }

    override fun onEvent(event: Event): Boolean {
        return false
    }

    override fun onGameLoop() {}
}
